ms_teams_importer: Support converting hosted content message attachment.

PieterCK opened the pull request:
**How changes were tested:**
Tested converting both using multiple processes (6) and only 1 process.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**



Images in message:
| MS Teams | Imported to Zulip |
|--------|--------|
| <img width="974" height="566" alt="image" src="https://github.com/user-attachments/assets/b7a53cd9-399b-4cba-b9b6-f39849efa351" />| <img width="1089" height="421" alt="image" src="https://github.com/user-attachments/assets/41b14588-7176-4393-bd65-29aab66837b1" />|
| <img width="787" height="339" alt="image" src="https://github.com/user-attachments/assets/a582de58-b70f-4a84-b555-c74ce831987c" /> | <img width="1089" height="357" alt="image" src="https://github.com/user-attachments/assets/b1c6bb7b-fee1-4b57-b211-352a968b776b" /> |

Custom emoji:
| MS Teams | Imported to Zulip |
|--------|--------|
| <img width="1092" height="308" alt="image" src="https://github.com/user-attachments/assets/0b510c97-90cd-4405-b95b-a2d9e2072a4c" /> | (Why: https://github.com/zulip/zulip/pull/37264#discussion_r2682755244) <img width="1104" height="114" alt="image" src="https://github.com/user-attachments/assets/bfcc77dc-bba0-4f72-a10c-753ccfc7e3e4" /> | 

Updated documentation:
- Added the "Hosted content" bullet point:
  <img width="1040" height="277" alt="image" src="https://github.com/user-attachments/assets/90c297e7-f8d9-44d2-9f41-18f367eac0f4" />

- Added a new required scope, `ChannelMessage.Read.All`:
  <img width="1040" height="413" alt="image" src="https://github.com/user-attachments/assets/fc125582-5900-4f1e-b0e6-930fc6c36752" />

- Added the explanation for `ChannelMessage.Read.All`:
  <img width="1021" height="508" alt="image" src="https://github.com/user-attachments/assets/d51d37da-1caa-4e9c-a4c2-7caa5fbbaa47" />

Fixes #37427.
<details>

<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
ms_teams_importer: Convert hosted content message attachments.

Fixes #37427.

PieterCK committed:
import_util: Include attachment url in AttachmentLinkResult.

PieterCK committed:
microsoft_teams: Don't call processes threads.

PieterCK committed:
ms_teams_importer: Return a MessageConversionResult instead of a tuple.

PieterCK committed:
ms_teams_importer: Tidy up process_messages's parameters.

This maintains alphabetical order and remove unecessary is_private
parameter.

timabbott commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -414,11 +432,127 @@ def is_microsoft_teams_event_message(message: MicrosoftTeamsFieldsT) -> bool:
+                created=os.path.getmtime(file_output_path),
+                size=os.path.getsize(file_output_path),
+                mimetype=raw_content_type,
+                name=os.path.basename(file_output_path),
```
I don't think we need to be overly concerned with that sort of micro-optimization here.

PieterCK committed:
import_util: Clarify UploadFileRequest's file path parameter.

This parameter doesn't expect the full file path, only the path id. See
the definition of path id `zerver/models/messages.py`.

PieterCK commented on starlight_help/src/content/docs/import-from-microsoft-teams.mdx:
```diff
@@ -157,6 +158,7 @@ tool:
 | [directory role list member](https://learn.microsoft.com/en-us/graph/api/directoryrole-list-members?view=graph-rest-1.0\&tabs=http) | Used to find users who are global administrators.      | `User.Read.All`                 |
 | [list directory role](https://learn.microsoft.com/en-us/graph/api/directoryrole-list?view=graph-rest-1.0\&tabs=http)                | Used to find the ID for the global administrator role. | `RoleManagement.Read.Directory` |
 | [list user](https://learn.microsoft.com/en-us/graph/api/user-list?view=graph-rest-1.0\&tabs=http)                                   | Used to find guest accounts.                           | `User.Read.All`                 |
+| [get hosted content](https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get?view=graph-rest-1.0\&tabs=http)       | Used to download hosted content attachments.           | `ChannelMessage.Read.All`       |
```
The least privileged permission for this API listed in [here](https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get?view=graph-rest-1.0&tabs=http) is actually `ChannelMessage.Read.Group`, but that permission doesn't exist:
<img width="1026" height="702" alt="image" src="https://github.com/user-attachments/assets/11581499-b047-4795-a7f8-e1887388aac4" />

So I use the higher level permission.

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -414,19 +442,128 @@ def is_microsoft_teams_event_message(message: MicrosoftTeamsFieldsT) -> bool:
+    "Hosted contents" are message file attachments that can be fetched straight
+    from Microsoft Graph API.
+
+    https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get
```
And the other type of file attachments is the Microsoft SharePoint attachment, they require a very different and more elaborate method to download. How they look in MS Teams:

Hosted contents:
<img width="1092" height="307" alt="image" src="https://github.com/user-attachments/assets/7b36b712-59d1-4c8f-94fe-d0c6c9f58bb9" />
<img width="1092" height="444" alt="image" src="https://github.com/user-attachments/assets/9859548c-2103-4ae4-a076-f1955cd3a3a7" />

MS SharePoint attachment:
<img width="1092" height="616" alt="image" src="https://github.com/user-attachments/assets/aec27207-d598-4a9b-a5a5-f396348181e8" />

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -414,19 +442,128 @@ def is_microsoft_teams_event_message(message: MicrosoftTeamsFieldsT) -> bool:
+    upload_records: list[UploadRecordData] = []
+
+    def export_file_and_get_zulip_url(match: regex.Match[str]) -> str:
+        # Hosted content attachments don't have a user-specified filename.
```
MS Teams also don't support Markdown text, so naming an image URL only done through alt text, which `convert_html_to_text` converts to this format `!({image alt text}]({graph api url})`. Saving this as a follow-up work to focus on the core uploads conversion first.

mateuszmandera commented on starlight_help/src/content/docs/import-from-microsoft-teams.mdx:
```diff
@@ -84,6 +83,9 @@ APIs](https://learn.microsoft.com/en-us/microsoftteams/migration-from-teams#step
 Some organization data are not exported by the Teams data export tool. The Zulip import
 tool will call several Microsoft Graph APIs to collect the following data:
 
+* [**Hosted contents**](https://learn.microsoft.com/en-us/graph/api/resources/chatmessagehostedcontent?view=graph-rest-1.0).
```
Above this there's the sentence `Some organization data are not exported by the Teams data export tool` - could you tweak that to `data is not ...`? I think that's more natural.

PieterCK replied:
```diff
diff --git a/starlight_help/src/content/docs/import-from-microsoft-teams.mdx b/starlight_help/src/content/docs/import-from-microsoft-teams.mdx
index 9697d36b84..c04ab48289 100644
--- a/starlight_help/src/content/docs/import-from-microsoft-teams.mdx
+++ b/starlight_help/src/content/docs/import-from-microsoft-teams.mdx
@@ -81,7 +81,7 @@ APIs](https://learn.microsoft.com/en-us/microsoftteams/migration-from-teams#step

 ### Authorize a Microsoft Graph API token

-Some organization data are not exported by the Teams data export tool. The Zulip import
+Some organization data is not exported by the Teams data export tool. The Zulip import
 tool will call several Microsoft Graph APIs to collect the following data:

 * [**Hosted contents**](https://learn.microsoft.com/en-us/graph/api/resources/chatmessagehostedcontent?view=graph-rest-1.0).
```

mateuszmandera commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -476,20 +620,42 @@ def process_messages(
+        if upload_records != []:
+            has_image = True
+            has_link = True
+            has_attachments = True
```
have you tested whether this causes any issues when a message has attachments that aren't images? since in such a case we're setting an incorrect value for `has_image`.

PieterCK replied:
Currently, hosted content only consists of image files, custom emoji, and code snippets. This currently only supports converting image files since custom emojis and code snippets are converted as custom HTML elements (e.g., `<customemoji>`), which `convert_html_to_text` converts into empty strings.

MS Teams has a different way of handling the conventional file attachments, those are stored in Microsoft Sharepoint, which involves a whole different process to handle.

The message payload with hosted content also don't give us useful metadata like the attachment's content type, so we only know that after the attachment is downloaded. Not sure if there's a clean implementation to compute accurate `has_image` for messages and still do the downloads in parallel, I suppose we could trade the parallel setup to get an accurate `has_image` property for each message?

I added a code comment to document this:
```diff
@@ -501,15 +510,27 @@ def process_hosted_content_attachments(
 ) -> AttachmentConversionResult:
     """
     "Hosted contents" are message file attachments that can be fetched straight
-    from Microsoft Graph API.
-
-    https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get
+    from Microsoft Graph API. As of April 2026, there are only 3 known types of
+    items that are categorized as "hosted contents":
+     * images
+     * custom emojis
+     * code snippets
+
+    Since custom emoji and code snippet are formatted as custom HTML blocks,
+    convert_html_to_text will turn them into empty strings, so this currently
+    only supports processing image hosted contents.
+
+    It's worth checking the API documentation for the hosted content once in a
+    while to verify what file types "hosted content" includes, since the code path
+    for this relies on that being only images:
+        https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get
     """
     upload_records: list[UploadRecordData] = []
```

mateuszmandera replied:
Ah, so do I understand correctly that e.g. custom emoji will **not** get appended to `upload_records` in `process_hosted_content_attachments`?

PieterCK replied:
yup, there's a case in `test_process_hosted_content_attachments` that verifies this.

timabbott replied:
We can merge this version as a checkpoint in this beta tool, but to me it seems like we should be aiming to compute `has_image` and friends from the rendering in the Zulip Markdown rendering pipeline, since that's where we definitively determine if there will be an image element in the Zulip output markdown, and that's what search fundamentally should use. 

@PieterCK can you look at doing that as your next task? It might be a pretty small patch.

PieterCK replied:
Looks like `markdown_convert` already does that if it has access to the message, but since `fix_message_rendered_content` only pass the message content, we're not doing that during import.

I included a fix for that and a couple other follow-ups in #39053. Still need some more polish before I can mark it as ready for review.

PieterCK replied:
Marked #39053 as ready for review, tagged Mateusz for review.

mateuszmandera commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -476,20 +620,42 @@ def process_messages(
+            zulip_message_id=message_id,
+        )
+        content = attachment_result.updated_content
+        upload_records += attachment_result.upload_records
```
I think causes a bug with the `if upload_records != []: has_image = True ....` logic. Since `upload_records` just adds more records as we iterate through messages, once we encounter the first message with attachments, all further messages will pass `upload_records != []:` and thus get these `has_*` flags set to True.

So that logic should be updated to only consider `attachment_result.upload_records`, not the accumulated `upload_records`. Calling this latter variable `accumulated_upload_records` should also be helpful in not getting these confused.

PieterCK replied:
Oof yeah, those variables having similar names must've tripped me there. Updated the variable names and added a test to catch messages with weird `has_*` conversion:

```diff
diff --git a/zerver/data_import/microsoft_teams.py b/zerver/data_import/microsoft_teams.py
index 2774173780..cfd3b276fe 100644
--- a/zerver/data_import/microsoft_teams.py
+++ b/zerver/data_import/microsoft_teams.py
@@ -569,7 +593,7 @@ def process_messages(
 ) -> MessageConversionResult:
     zerver_usermessage: list[ZerverFieldsT] = []
     zerver_messages: list[ZerverFieldsT] = []
-    upload_records: list[UploadRecordData] = []
+    accumulated_batch_upload_records: list[UploadRecordData] = []

     for message in messages:
         if is_microsoft_teams_event_message(message):
@@ -632,9 +656,11 @@ def process_messages(
             zulip_message_id=message_id,
         )
         content = attachment_result.updated_content
-        upload_records += attachment_result.upload_records
+        accumulated_batch_upload_records += attachment_result.upload_records

-        if upload_records != []:
+        if attachment_result.upload_records != []:
+            # Currently process_hosted_content_attachments only supports processeing
+            # images so has_image is always true in this case.
             has_image = True
             has_link = True
             has_attachments = True
@@ -678,7 +704,7 @@ def process_messages(
     return MessageConversionResult(
         zerver_messages=zerver_messages,
         zerver_usermessages=zerver_usermessage,
-        upload_records=upload_records,
+        upload_records=accumulated_batch_upload_records,
     )


@@ -742,7 +768,7 @@ def convert_messages(
                 team_id=team_id,
             )
     total_attachment_records: list[AttachmentRecordData] = []
-    total_upload_records: list[UploadRecordData] = []
+    total_accumulated_upload_records: list[UploadRecordData] = []
     with run_parallel_queue(
         partial(
             download_and_export_microsoft_teams_upload_file, total_attachment_records, output_dir
@@ -766,7 +792,7 @@ def convert_messages(
                 realm_id=realm_id,
                 subscriber_map=subscriber_map,
             )
-            total_upload_records += conversion_result.upload_records
+            total_accumulated_upload_records += conversion_result.upload_records
             create_converted_data_files(
                 dict(
                     zerver_message=conversion_result.zerver_messages,
@@ -776,7 +802,7 @@ def convert_messages(
                 f"/messages-{dump_file_id:06}.json",
             )
             dump_file_id += 1
-    return (total_upload_records, total_attachment_records)
+    return (total_accumulated_upload_records, total_attachment_records)


 def do_convert_directory(
diff --git a/zerver/tests/test_microsoft_teams_importer.py b/zerver/tests/test_microsoft_teams_importer.py
index 7e2250ad2c..4e60ad44ed 100644
--- a/zerver/tests/test_microsoft_teams_importer.py
+++ b/zerver/tests/test_microsoft_teams_importer.py
@@ -467,17 +468,33 @@ class MicrosoftTeamsImporterIntegrationTest(MicrosoftTeamsImportTestCase):

         imported_attachments = Attachment.objects.filter(realm=test_realm)
         self.assertTrue(imported_attachments.exists())
+        message_ids_with_attachment = set()
         for attachment in imported_attachments:
             for message_with_attachment in Message.objects.filter(
                 realm=test_realm, attachment=attachment
             ):
                 self.assertTrue(message_with_attachment.has_attachment)
-                self.assertTrue(message_with_attachment.has_image)
+                self.assertEqual(
+                    message_with_attachment.has_image,
+                    attachment.content_type in THUMBNAIL_ACCEPT_IMAGE_TYPES,
+                )
                 self.assertTrue(message_with_attachment.has_link)
                 assert isinstance(message_with_attachment.rendered_content, str)
                 self.assertIn(
                     f"/user_uploads/{attachment.path_id}", message_with_attachment.rendered_content
                 )
+                message_ids_with_attachment.add(message_with_attachment.id)
+
+        # Make sure no weird message conversion with wrong has_attachment, has_link
+        # and has_image property.
+        self.assertSetEqual(
+            message_ids_with_attachment,
+            set(
+                Message.objects.filter(
+                    realm=test_realm, has_link=True, has_attachment=True, has_image=True
+                ).values_list("id", flat=True)
+            ),
+        )


 class MicrosoftTeamsImporterUnitTest(MicrosoftTeamsImportTestCase):
```

mateuszmandera reviewed (commented):
Sorry it took so long for me to get to reviewing this! Posted a couple of comments. The changes look solid besides that. I didn't look at tests in detail yet.

PieterCK committed:
markdown_util: Support creating Markdown syntax with inline thumbnmail.

This makes get_markdown_link_for_url be able to output "![name](url)"
syntax for image urls.

PieterCK commented:
Changes:
- Fixed documentation https://github.com/zulip/zulip/pull/37264#discussion_r3094594825
- Added documentation to `process_hosted_content_attachments` https://github.com/zulip/zulip/pull/37264#discussion_r3094728392
- Fixed a bug generating messages with wrong `has_*` property https://github.com/zulip/zulip/pull/37264#discussion_r3094766143
- Updated the regex pattern and logic for hosted content url to be more robust
- Added better conversion of image attachment in message content, now they use inline thumbnail and the filename is sanitized.
- Minor nits like added a code comment in a test.

<details>

<summary>
git diff from previous version
</summary>

```diff
diff --git a/starlight_help/src/content/docs/import-from-microsoft-teams.mdx b/starlight_help/src/content/docs/import-from-microsoft-teams.mdx
index 9697d36b84..c04ab48289 100644
--- a/starlight_help/src/content/docs/import-from-microsoft-teams.mdx
+++ b/starlight_help/src/content/docs/import-from-microsoft-teams.mdx
@@ -81,7 +81,7 @@ APIs](https://learn.microsoft.com/en-us/microsoftteams/migration-from-teams#step

 ### Authorize a Microsoft Graph API token

-Some organization data are not exported by the Teams data export tool. The Zulip import
+Some organization data is not exported by the Teams data export tool. The Zulip import
 tool will call several Microsoft Graph APIs to collect the following data:

 * [**Hosted contents**](https://learn.microsoft.com/en-us/graph/api/resources/chatmessagehostedcontent?view=graph-rest-1.0).
diff --git a/zerver/data_import/microsoft_teams.py b/zerver/data_import/microsoft_teams.py
index 2774173780..cfd3b276fe 100644
--- a/zerver/data_import/microsoft_teams.py
+++ b/zerver/data_import/microsoft_teams.py
@@ -1,5 +1,6 @@
 import logging
 import os
+import re
 import shutil
 from collections import defaultdict
 from collections.abc import Callable, Iterable
@@ -37,6 +38,7 @@ from zerver.data_import.import_util import (
 )
 from zerver.data_import.sequencer import NEXT_ID
 from zerver.lib.export import MESSAGE_BATCH_CHUNK_SIZE, do_common_export_processes
+from zerver.lib.markdown import get_markdown_link_for_url
 from zerver.lib.mime_types import bare_content_type
 from zerver.lib.parallel import run_parallel_queue
 from zerver.lib.partial import partial
@@ -207,7 +209,14 @@ class ODataQueryParameter:
 MICROSOFT_GRAPH_API_URL = "https://graph.microsoft.com/v1.0{endpoint}"

 # https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get
-GRAPH_API_HOSTED_CONTENT_REGEX = r"https://graph\.microsoft\.com/v1\.0/teams/[^/]+/channels/[^/]+/messages/[^/]+/hostedContents/[^/]+/\$value"
+HOSTED_CONTENT_GRAPH_API_URL_REGEX = r"https://graph\.microsoft\.com/v1\.0/teams/[^/]+/channels/[^/]+/messages/[^/]+/hostedContents/[^/]+/\$value"
+HOSTED_CONTENT_FILE_REGEX = rf"""
+            \[
+               (?P<file_name>[^\]]+)
+            \]\(
+               (?P<api_url>{HOSTED_CONTENT_GRAPH_API_URL_REGEX})
+            \)
+            """


 def get_microsoft_graph_api_data(
@@ -501,15 +510,27 @@ def process_hosted_content_attachments(
 ) -> AttachmentConversionResult:
     """
     "Hosted contents" are message file attachments that can be fetched straight
-    from Microsoft Graph API.
-
-    https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get
+    from Microsoft Graph API. As of April 2026, there are only 3 known types of
+    items that are categorized as "hosted contents":
+     * images
+     * custom emojis
+     * code snippets
+
+    Since custom emoji and code snippet are formatted as custom HTML blocks,
+    convert_html_to_text will turn them into empty strings, so this currently
+    only supports processing image hosted contents.
+
+    It's worth checking the API documentation for the hosted content once in a
+    while to verify what file types "hosted content" includes, since the code path
+    for this relies on that being only images:
+        https://learn.microsoft.com/en-us/graph/api/chatmessagehostedcontent-get
     """
     upload_records: list[UploadRecordData] = []

     def export_file_and_get_zulip_url(match: regex.Match[str]) -> str:
-        # Hosted content attachments don't have a user-specified filename.
-        attachment_data = get_attachment_path_and_content("", teams_message_id, realm_id)
+        file_name = match.group("file_name")
+        hosted_content_api_url = match.group("api_url")
+        attachment_data = get_attachment_path_and_content(file_name, teams_message_id, realm_id)

         upload_records.append(
             UploadRecordData(
@@ -530,14 +551,16 @@ def process_hosted_content_attachments(
                 sender_id=zulip_user_id,
                 upload_file_request=UploadFileRequest(
                     output_file_path_id=attachment_data.path_id,
-                    request_url=match.group(),
+                    request_url=hosted_content_api_url,
                     params=None,
                     headers={"Authorization": f"Bearer {microsoft_graph_api_token}"},
                     kwargs={},
                 ),
             )
         )
-        return attachment_data.url
+        # Use inline thumbnail here since MS Teams only supports customizing the hosted
+        # content image's alt tag.
+        return get_markdown_link_for_url(file_name, attachment_data.url, inline_thumbnail=True)

     if is_direct_message_type:
         # TODO: hosted content URL for private chats have a different format, we
@@ -545,9 +568,10 @@ def process_hosted_content_attachments(
         return AttachmentConversionResult(updated_content=content, upload_records=upload_records)
     else:
         updated_content = regex.sub(
-            GRAPH_API_HOSTED_CONTENT_REGEX,
+            HOSTED_CONTENT_FILE_REGEX,
             export_file_and_get_zulip_url,
             content,
+            flags=re.VERBOSE,
         )
         return AttachmentConversionResult(
             updated_content=updated_content,
@@ -569,7 +593,7 @@ def process_messages(
 ) -> MessageConversionResult:
     zerver_usermessage: list[ZerverFieldsT] = []
     zerver_messages: list[ZerverFieldsT] = []
-    upload_records: list[UploadRecordData] = []
+    accumulated_batch_upload_records: list[UploadRecordData] = []

     for message in messages:
         if is_microsoft_teams_event_message(message):
@@ -632,9 +656,11 @@ def process_messages(
             zulip_message_id=message_id,
         )
         content = attachment_result.updated_content
-        upload_records += attachment_result.upload_records
+        accumulated_batch_upload_records += attachment_result.upload_records

-        if upload_records != []:
+        if attachment_result.upload_records != []:
+            # Currently process_hosted_content_attachments only supports processeing
+            # images so has_image is always true in this case.
             has_image = True
             has_link = True
             has_attachments = True
@@ -678,7 +704,7 @@ def process_messages(
     return MessageConversionResult(
         zerver_messages=zerver_messages,
         zerver_usermessages=zerver_usermessage,
-        upload_records=upload_records,
+        upload_records=accumulated_batch_upload_records,
     )


@@ -742,7 +768,7 @@ def convert_messages(
                 team_id=team_id,
             )
     total_attachment_records: list[AttachmentRecordData] = []
-    total_upload_records: list[UploadRecordData] = []
+    total_accumulated_upload_records: list[UploadRecordData] = []
     with run_parallel_queue(
         partial(
             download_and_export_microsoft_teams_upload_file, total_attachment_records, output_dir
@@ -766,7 +792,7 @@ def convert_messages(
                 realm_id=realm_id,
                 subscriber_map=subscriber_map,
             )
-            total_upload_records += conversion_result.upload_records
+            total_accumulated_upload_records += conversion_result.upload_records
             create_converted_data_files(
                 dict(
                     zerver_message=conversion_result.zerver_messages,
@@ -776,7 +802,7 @@ def convert_messages(
                 f"/messages-{dump_file_id:06}.json",
             )
             dump_file_id += 1
-    return (total_upload_records, total_attachment_records)
+    return (total_accumulated_upload_records, total_attachment_records)


 def do_convert_directory(
diff --git a/zerver/lib/markdown/__init__.py b/zerver/lib/markdown/__init__.py
index cb64be5c09..686eaccdf1 100644
--- a/zerver/lib/markdown/__init__.py
+++ b/zerver/lib/markdown/__init__.py
@@ -2832,9 +2832,12 @@ def render_message_markdown(
     return rendering_result


-def get_markdown_link_for_url(filename: str, url: str) -> str:
+def get_markdown_link_for_url(filename: str, url: str, inline_thumbnail: bool = False) -> str:
     # Our markdown has no escaping, so we cannot link any
     # text containing brackets; strip them from the
     # filename we're linking.
     filename = re.sub(r"\[|\]", "", filename)
-    return f"[{filename}]({url})"
+    markdown_link = f"[{filename}]({url})"
+    if inline_thumbnail:
+        markdown_link = "!" + markdown_link
+    return markdown_link
diff --git a/zerver/tests/test_microsoft_teams_importer.py b/zerver/tests/test_microsoft_teams_importer.py
index 7e2250ad2c..4e60ad44ed 100644
--- a/zerver/tests/test_microsoft_teams_importer.py
+++ b/zerver/tests/test_microsoft_teams_importer.py
@@ -18,7 +18,7 @@ from typing_extensions import ParamSpec

 from zerver.data_import.import_util import AttachmentRecordData, convert_html_to_text, get_data_file
 from zerver.data_import.microsoft_teams import (
-    GRAPH_API_HOSTED_CONTENT_REGEX,
+    HOSTED_CONTENT_GRAPH_API_URL_REGEX,
     MICROSOFT_GRAPH_API_URL,
     ChannelMetadata,
     MicrosoftTeamsFieldsT,
@@ -41,6 +41,7 @@ from zerver.lib.import_realm import do_import_realm
 from zerver.lib.partial import partial
 from zerver.lib.test_classes import ZulipTestCase
 from zerver.lib.test_helpers import read_test_image_file
+from zerver.lib.thumbnail import THUMBNAIL_ACCEPT_IMAGE_TYPES
 from zerver.lib.topic import messages_for_topic
 from zerver.models.messages import Attachment, Message
 from zerver.models.realms import Realm, get_realm
@@ -228,7 +229,7 @@ def mock_microsoft_graph_api_calls(
         )
         responses.add(
             responses.GET,
-            re.compile(GRAPH_API_HOSTED_CONTENT_REGEX),
+            re.compile(HOSTED_CONTENT_GRAPH_API_URL_REGEX),
             body=read_test_image_file("img.png"),
             content_type="image/png",
             status=200,
@@ -467,17 +468,33 @@ class MicrosoftTeamsImporterIntegrationTest(MicrosoftTeamsImportTestCase):

         imported_attachments = Attachment.objects.filter(realm=test_realm)
         self.assertTrue(imported_attachments.exists())
+        message_ids_with_attachment = set()
         for attachment in imported_attachments:
             for message_with_attachment in Message.objects.filter(
                 realm=test_realm, attachment=attachment
             ):
                 self.assertTrue(message_with_attachment.has_attachment)
-                self.assertTrue(message_with_attachment.has_image)
+                self.assertEqual(
+                    message_with_attachment.has_image,
+                    attachment.content_type in THUMBNAIL_ACCEPT_IMAGE_TYPES,
+                )
                 self.assertTrue(message_with_attachment.has_link)
                 assert isinstance(message_with_attachment.rendered_content, str)
                 self.assertIn(
                     f"/user_uploads/{attachment.path_id}", message_with_attachment.rendered_content
                 )
+                message_ids_with_attachment.add(message_with_attachment.id)
+
+        # Make sure no weird message conversion with wrong has_attachment, has_link
+        # and has_image property.
+        self.assertSetEqual(
+            message_ids_with_attachment,
+            set(
+                Message.objects.filter(
+                    realm=test_realm, has_link=True, has_attachment=True, has_image=True
+                ).values_list("id", flat=True)
+            ),
+        )


 class MicrosoftTeamsImporterUnitTest(MicrosoftTeamsImportTestCase):
@@ -718,7 +735,7 @@ class MicrosoftTeamsImporterUnitTest(MicrosoftTeamsImportTestCase):
     def test_process_hosted_content_attachments(self) -> None:
         responses.add(
             responses.GET,
-            re.compile(GRAPH_API_HOSTED_CONTENT_REGEX),
+            re.compile(HOSTED_CONTENT_GRAPH_API_URL_REGEX),
             body=read_test_image_file("img.png"),
             content_type="image/png",
             status=200,
@@ -844,6 +861,8 @@ class MicrosoftTeamsImporterUnitTest(MicrosoftTeamsImportTestCase):
                 # Increment the fixture IDs to make them unique.
                 fixture.teams_message_id += str(id_count)
                 fixture.zulip_message_id += id_count
+                # Multiply by 10 to help verify that we're computing each
+                # object ID correctly for this code path.
                 fixture.zulip_user_id += id_count * 10

                 total_attachment_records: list[AttachmentRecordData] = []
```
</details>

PieterCK commented:
hmm the CI seems to be especially flakey right now. Will mark this as ready for review once that's cleared up

PieterCK commented:
@mateuszmandera -- thanks for reviewing this! I think this is ready for review again.

I think the main issue is the potentially inaccurate `has_image` property we’re setting if the hosted content API includes file types other than images in the future (https://github.com/zulip/zulip/pull/37264#discussion_r3094728392). I also haven’t fully investigated which code paths the `has_image` flag affects, so I’m not sure how severe the bug would be if messages have an incorrect `has_image` property.

mateuszmandera commented:
okay, i think that's probably solid enough for this importer for `integration review`. @timabbott Can you take a look next?

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2832,9 +2832,12 @@ def render_message_markdown(
 
 
-def get_markdown_link_for_url(filename: str, url: str) -> str:
+def get_markdown_link_for_url(filename: str, url: str, inline_thumbnail: bool = False) -> str:
```
`inline_thumbnail` seems like the wrong name. I think a parallel `get_markdown_image_for_url` function would be a better long-term abstraction, since it is a "Markdown image element", not a "thumbnail", at the Markdown spec level.

PieterCK replied:
Added a commit for this in #39053

timabbott commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -476,20 +644,44 @@ def process_messages(
+            has_attachments = True
+        else:
+            has_image = False
+            has_link = False
```
`has_link` here is wrong -- that should be checking if the result has links. But given the end-to-end proposal for where these fields should be computed that is discussed elsewhere, probably the fix is the same.

PieterCK replied:
This also gets fixed by #39053.

timabbott commented on zerver/tests/test_microsoft_teams_importer.py:
```diff
@@ -680,3 +730,186 @@ def test_get_batched_export_message_data(self) -> None:
+                            rf"^{realm_id}/.*/.*/{fixture.teams_message_id}$",
+                        )
+        finally:
+            shutil.rmtree(output_dir)
```
Do these tests need a `rmtree` in a `finally` clause, or does the test class teardown method take care of it?

PieterCK replied:
The doc string for `tempfile.mkdtemp` does say:

> Caller is responsible for deleting the directory when done with it.

And `tearDown` looks like it only cleans up database transactions:

```python
def tearDownClass(cls):
        if (
            cls._databases_support_transactions()
            and cls._databases_support_savepoints()
        ):
            cls._rollback_atomics(cls.cls_atomics)
            for conn in connections.all(initialized_only=True):
                conn.close()
        super().tearDownClass()
```

timabbott commented:
Seems like a great checkpoint, merged, thanks @PieterCK and @mateuszmandera!

I posted a few comments on follow-ups -- the `has_image` calculation using the Markdown processor is most timely, since it'd be nice if we can do that before end users actually use this importer.
